python/perf-kernels/flash-attention-seqlen-padded.py

```diff
@@ -0,0 +1,772 @@
+        block_shape=(BLOCK_DMODEL, BLOCK_N),
+        order=(0, 1)
+    )
+    DO_block_ptr = tl.make_block_ptr(
+        base=DO + qvk_offset,
+        shape=(N_CTX, BLOCK_DMODEL),
+        strides=(stride_qm, stride_qk),
+        offsets=(start_m * BLOCK_M, 0),
+        block_shape=(BLOCK_M, BLOCK_DMODEL),
+        order=(1, 0)
+    )
+    # pointer to row-wise quantities in value-like data
+    D_ptrs = D + off_hz * N_CTX
+    l_ptrs = L + off_hz * N_CTX
+    qk_scale = sm_scale * 1.44269504
+    # load q and do: they will stay in SRAM throughout
+    q = tl.load(Q_block_ptr)
+    q = (q * qk_scale).to(q.dtype)
+    do = tl.load(DO_block_ptr)
+    Di = tl.load(D_ptrs + offs_m)
+    l_i = tl.load(l_ptrs + offs_m)
+    dq = tl.zeros([BLOCK_M, BLOCK_DMODEL], dtype=tl.float32)
+    # loop over k, v
+    lo = 0
+    hi = (start_m + 1) * BLOCK_M
+    for start_n in range(lo, hi, BLOCK_N):
+        # -- load k, v --
+        k = tl.load(K_block_ptr)
+        v = tl.load(V_block_ptr)
+        # -- compute qk ----
+        qk = tl.dot(q, k)
+        qk = tl.where(offs_m[:, None] >= (offs_n[None, :] + start_n), qk, float("-inf"))
+        p = tl.math.exp2(qk - l_i[:, None])
+        # compute dp = dot(v, do)
+        dp = tl.zeros([BLOCK_M, BLOCK_N], dtype=tl.float32) - Di[:, None]
+        dp += tl.dot(do, v)
+        # compute ds = p * (dp - delta[:, None])
+        ds = p * dp
+        # compute dq. Unfortunately we cannot avoid transpose here as this loop
+        # uses k both normal and transpose.
+        dq += tl.dot(ds.to(Q.dtype.element_ty), tl.trans(k))
+        # update pointers
+        K_block_ptr = tl.advance(K_block_ptr, (0, BLOCK_N))
+        V_block_ptr = tl.advance(V_block_ptr, (0, BLOCK_N))
+    # initialize pointers to output
+    DQ_block_ptr = tl.make_block_ptr(
+        base=DQ + qvk_offset,
+        shape=(N_CTX, BLOCK_DMODEL),
+        strides=(stride_qm, stride_qk),
+        offsets=(start_m * BLOCK_M, 0),
+        block_shape=(BLOCK_M, BLOCK_DMODEL),
+        order=(1, 0)
+    )
+    tl.store(DQ_block_ptr, (dq * sm_scale).to(tl.float16))
+
+empty = torch.empty(128, device="cuda")
+
+
+class _attention(torch.autograd.Function):
+    @staticmethod
+    def forward(ctx, q, k, v, causal, sm_scale, split_kernel=False):
+        # shape constraints
+        _, _, seqlen, Lq = q.shape
+        Lk, Lv = k.shape[-1], v.shape[-1]
+        assert Lq == Lk and Lk == Lv
+        assert Lk in {16, 32, 64, 128}
+        # For now we assume K and V seqlen = Q seqlen
+        assert seqlen == k.shape[-2] and seqlen == v.shape[-2]
+
+        # We've derived these previously from tuning the kernel
+        BLOCK_M = 256 if Lq == 128 else 128
+        BLOCK_N = 64 #128 if Lq == 128 else 64
+        waves_per_eu = 2 if Lq == 128 else 3
+        num_warps = 8 if Lq == 128 else 4
+        pre_load_v = False if Lq == 128 else True
+        
+        stage = 3 if causal else 1
+        need_padding = True if seqlen % BLOCK_M else False
```

Comment:
I think this should also be BLOCK_N

Reply:
maybe this is the reason for out of memory error.

Reply:
We don't have to add the same padding for M and N. do we?

Reply:
We don't. Here is what I understand:

1) Block Q: This size is always BLOCK_M x BLOCK_DMODEL as defined by the make_block_ptr() call, and the boundary check takes care of how much padding needs to be added. We don't need to worry about this.

2) Block K/V: This block is always BLOCK_DMODEL x BLOCK_N. As with Q, we use boundary check to ensure we do not overflow.

3) QK result: This is BLOCK_M x BLOCK_N. Both M and N directions have padding. The amount of padding in M is irrelevant, because M dim here is also M dim of output, and we use boundary check to ensure we do not overflow. But the N dim is critical as it is K dim of PV GEMM. We do not store P, so we cannot use boundary check. Instead we use tl.where, which requires the number of padded tokens we have in the N dim.

So we need to know two things (a) if we will need padding and (b) how much padding we will need. 

(a) should use the larger of M and N, since the larger of pow2 is a multiple of smaller but the converse is not true, of course.

(b) should use the dimension that we will use in the tl.where() call, which is N, so we need to know how much we padded in N dim.

(a) is why need_padding uses mod M. (b) is why extra_tokens uses mod N. Does this make sense?

Reply:
> so I was thinking for the current config (M = 256, N = 64 / 128), it should be BLOCK_M because, consider sequence length of 192. 192 % 64 == 0 but we will still need padding in the M direction

Lets's go back to this simple example. 
The padding along the M (Q tensor) direction is handled automatically by load(Q_block_ptr, boundary_check) stuff. It doesn't need `need_padding` to tell it to do so.

Reply:
> The padding along the M (Q tensor) direction is handled automatically by load(Q_block_ptr, boundary_check) stuff. It doesn't need need_padding to tell it to do so.

This is also true for N. It is automatically handled by load(K_block_ptr, boundary_check).

Reply:
But you need `need_padding` to control the last loop.

Reply:
Okay, I see what you mean. need_padding controls that last loop, which depends on whether K needs to be padded in the N dimension. Will fix. Thanks for explaining.

Reply:
Fixed.